tests/kernel/smp/src/main.c

```diff
@@ -624,6 +630,573 @@ void test_smp_ipi(void)
+			"lock the mutex failed");
+	printk("(cpu%d)after wait for mutex lock\n", arch_curr_cpu()->id);
+
+	irq_unlock(key);
+	printk("(cpu%d)t1 after unlock\n", arch_curr_cpu()->id);
+
+	k_mutex_unlock(&smutex);
+	printk("(cpu%d)after mutex unlock\n", arch_curr_cpu()->id);
+}
+
+/**
+ * @brief Test the thread lock mutex after locking the interrupt
+ *
+ * @ingroup kernel_smp_tests
+ *
+ * @details Validate if the behavior is correct, while the thread
+ * lock mutex after locking the interrupt.
+ */
+void test_smp_irq_lock_mutex(void)
+{
+	int key;
+
+	k_mutex_init(&smutex);
+
+	tinfo[0].tid =
+	k_thread_create(&tthread[0], tstack[0], STACK_SIZE,
+			(k_thread_entry_t)t_try_lock_mutex,
+			&smutex, NULL, NULL,
+			K_PRIO_PREEMPT(0),
+			K_INHERIT_PERMS, K_NO_WAIT);
+
+	k_mutex_lock(&smutex, K_FOREVER);
+	printk("(cpu%d)t0 hold the mutex\n", arch_curr_cpu()->id);
+
+	key = irq_lock();
+	printk("(cpu%d)t0 hold the lock\n", arch_curr_cpu()->id);
+
+	k_mutex_unlock(&smutex);
+	printk("(cpu%d)t0 after mutex unlock\n", arch_curr_cpu()->id);
+
+	k_mutex_lock(&smutex, K_FOREVER);
+	printk("(cpu%d)t0 after mutex lock again\n", arch_curr_cpu()->id);
+
+	irq_unlock(key);
+	printk("(cpu%d)t0 after unlock\n", arch_curr_cpu()->id);
+
+	k_thread_join(tinfo[0].tid, K_FOREVER);
+	cleanup_resources();
+}
+#endif
+
+static void t1_mutex_lock(void *p1, void *p2, void *p3)
+{
+	/* t1 will get mutex first */
+	k_mutex_lock((struct k_mutex *)p1, K_FOREVER);
+
+	k_msleep(2);
+
+	k_mutex_unlock((struct k_mutex *)p1);
+}
+
+static void t2_mutex_lock(void *p1, void *p2, void *p3)
+{
+	k_mutex_lock((struct k_mutex *)p1, K_FOREVER);
+
+	k_mutex_unlock((struct k_mutex *)p1);
+}
+
+static void t2_mutex_lock_with_irq(void *p1, void *p2, void *p3)
+{
+	int key = irq_lock();
+
+	k_mutex_lock((struct k_mutex *)p1, K_FOREVER);
+
+	k_mutex_unlock((struct k_mutex *)p1);
+	irq_unlock(key);
+}
+
+/**
+ * @brief Test scenairo that a thread release the global lock
+ *
+ * @ingroup kernel_smp_tests
+ *
+ * @details Validate the scenario that make the internal APIs of SMP
+ * z_smp_release_global_lock() to be called.
+ */
+void test_smp_release_global_lock(void)
+{
+	k_mutex_init(&smutex);
+
+	tinfo[0].tid =
+	k_thread_create(&tthread[0], tstack[0], STACK_SIZE,
+			(k_thread_entry_t)t1_mutex_lock,
+			&smutex, NULL, NULL,
+			K_PRIO_PREEMPT(5),
+			K_INHERIT_PERMS, K_NO_WAIT);
+
+	tinfo[1].tid =
+	k_thread_create(&tthread[1], tstack[1], STACK_SIZE,
+		(k_thread_entry_t)t2_mutex_lock,
+			&smutex, NULL, NULL,
+			K_PRIO_PREEMPT(3),
+			K_INHERIT_PERMS, K_MSEC(1));
+
+	/* hold the cpu */
+	k_busy_wait(20000);
+
+	k_thread_join(tinfo[1].tid, K_FOREVER);
+	k_thread_join(tinfo[0].tid, K_FOREVER);
+	cleanup_resources();
+}
+
+/**
+ * @brief Test scenairo that a thread release the global lock
+ *
+ * @ingroup kernel_smp_tests
+ *
+ * @details Validate the scenario that make the internal APIs of SMP
+ * z_smp_release_global_lock() to be called.
+ */
+void test_smp_release_global_lock_irq(void)
+{
```

Comment:
The test case will always succeed. How do we know z_smp_global_lock() has been called.

Same to test_smp_release_global_lock().

Reply:
@LeiW000 Yes, you are right. There should be some checkpoint to see if the condition meets or not. Let me try to refine it.
Thanks for pointing out this!